### PR TITLE
SocketsHttpHandler: fix logic to check for Proxy-support header

### LINKED ISSUE #27872: SocketHttpHandler is failing NTLM authentication to proxy
I essentially use example from #27870 When running on Windows using WinHTTPHandler it works just fine. On Linux I did verify my proxy setup using curl directly. 

Note, that server expects NTLM directly instead of Negotiate/NTLM

> HTTP/1.1 407 Proxy Authentication Required
> Server: squid/3.5.12
> Mime-Version: 1.0
> Date: Thu, 08 Mar 2018 21:26:30 GMT
> Content-Type: text/html;charset=utf-8
> Content-Length: 3566
> X-Squid-Error: ERR_CACHE_ACCESS_DENIED 0
> Vary: Accept-Language
> Content-Language: en
> Proxy-Authenticate: NTLM
> X-Cache: MISS from CLRPERFTST003
> X-Cache-Lookup: NONE from CLRPERFTST003:3128
> Via: 1.1 CLRPERFTST003 (squid/3.5.12)
> Connection: keep-alive
> 

On Linux I tried to force NTLM in TryGetValidAuthenticationChallengeForScheme() and CheckIfProxySupportsConnectionAuth() and I got following trace.

> System.ComponentModel.Win32Exception (0x80090020): GSSAPI operation failed with error - An invalid status code was supplied (Unknown error).
>    at System.Net.Security.NegotiateStreamPal.AcquireCredentialsHandle(String package, Boolean isServer, NetworkCredential credential) in /home/clruser/git/corefx/src/Common/src/System/Net/Security/NegotiateStreamPal.Unix.cs:line 313
>    at System.Net.NTAuthentication.Initialize(Boolean isServer, String package, NetworkCredential credential, String spn, ContextFlagsPal requestedContextFlags, ChannelBinding channelBinding) in /home/clruser/git/corefx/src/Common/src/System/Net/NTAuthentication.Common.cs:line 127
>    at System.Net.NTAuthentication..ctor(Boolean isServer, String package, NetworkCredential credential, String spn, ContextFlagsPal requestedContextFlags, ChannelBinding channelBinding) in /home/clruser/git/corefx/src/Common/src/System/Net/NTAuthentication.Common.cs:line 98
>    at System.Net.Http.AuthenticationHelper.SendWithNtAuthAsync(HttpRequestMessage request, Uri authUri, ICredentials credentials, Boolean isProxyAuth, HttpConnection connection, CancellationToken cancellationToken) in /home/clruser/git/corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs:line 57
>    at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in /home/clruser/git/corefx/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:line 255

That was somewhat surprising as I did not have Kerberos configured and curl works fine. 
On Windows I did not dig deeper but I don't see any response to initial 407.

### PR DESCRIPTION
**Pull request by geoffkizer**
Currently we're looking for this header when we do proxy auth.  This is wrong.  We should look for this header when we are trying to do regular auth through a proxy.

@davidsh @wfurt @stephentoub 

Fixes #27872 

@wfurt can you test and confirm that this fixes #27872?  (Windows only)

### CODE REVIEW

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs [hunk 1]
```diff
@@ -41,8 +41,13 @@ private static bool CheckIfProxySupportsConnectionAuth(HttpResponseMessage respo
         {
             HttpResponseMessage response = await InnerSendAsync(request, isProxyAuth, connection, cancellationToken).ConfigureAwait(false);
 
-            if (TryGetAuthenticationChallenge(response, isProxyAuth, authUri, credentials, out AuthenticationChallenge challenge) &&
-                (!isProxyAuth || CheckIfProxySupportsConnectionAuth(response)))
+            if (!isProxyAuth && connection.UsingProxy && !CheckIfProxySupportsConnectionAuth(response))
```
[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs]: Could `CheckIfProxySupportsConnectionAuth` just be `ProxySupportsConnectionAuth`?

[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs]: Yeah, that's better, I'll change.

[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs]: Re isProxyAuth:

This routine (and the similar one in AuthenticationHelper for basic/digest) is used in two cases:
(1) Authenticate against proxy (isProxyAuth = true)
(2) Authenticate against origin server (isProxyAuth = false)

The logic is basically the same between these two cases, just different status code (407 vs 401) and different header names (Proxy-Authenticate vs Www-Authenticate, Proxy-Authorization vs Authorization).  So we just use the isProxyAuth flag to distinguish.

For example, this helper is called:

```csharp
        private static HttpHeaderValueCollection<AuthenticationHeaderValue> GetResponseAuthenticationHeaderValues(HttpResponseMessage response, bool isProxyAuth)
        {
            return isProxyAuth ?
                response.Headers.ProxyAuthenticate :
                response.Headers.WwwAuthenticate;
        }
```

I'm open to a better parameter name than "isProxyAuth" if you have suggestions...

[the remark this answers is not in the document]
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs]: Thanks for the explanation.  Maybe just a comment on the parameter explaining what it is.

[the remark this answers is not in the document]
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/AuthenticationHelper.NtAuth.cs]: perhaps "proxyAuthMode"?

**Comment by geoffkizer** [comment]
@wfurt, I'm going to merge.  Let me know when you have a chance to try this in your setup.

**Pull request merged**